haystack/telemetry.py

```diff
@@ -0,0 +1,343 @@
+    """
+    os.environ[HAYSTACK_TELEMETRY_LOGGING_TO_FILE_ENABLED] = "False"
+    logger.info(f"Writing events to log file {LOG_PATH} has been disabled.")
+
+
+def is_telemetry_enabled() -> bool:
+    """
+    Returns False if telemetry is disabled via an environment variable, otherwise True.
+    """
+    telemetry_environ = os.environ.get(HAYSTACK_TELEMETRY_ENABLED, "True")
+    return telemetry_environ.lower() != "false"
+
+
+def is_telemetry_logging_to_file_enabled() -> bool:
+    """
+    Returns False if logging telemetry events to a file is disabled via an environment variable, otherwise True.
+    """
+    telemetry_environ = os.environ.get(HAYSTACK_TELEMETRY_LOGGING_TO_FILE_ENABLED, "False")
+    return telemetry_environ.lower() != "false"
+
+
+def send_event_if_public_demo(func):
+    """
+    Can be used as a decorator to send an event only if HAYSTACK_EXECUTION_CONTEXT is "public_demo"
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        exec_context = os.environ.get(HAYSTACK_EXECUTION_CONTEXT, "")
+        if exec_context == "public_demo":
+            send_custom_event(event="demo query executed", payload=kwargs)
+        return func(*args, **kwargs)
+
+    return wrapper
+
+
+def send_event(func):
+    """
+    Can be used as a decorator to send an event formatted like 'Pipeline.eval executed'
+    with additional parameters as defined in TrackedParameters ('add_isolated_node_eval') and
+    metadata, such as os_version
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        send_custom_event(event=f"{type(args[0]).__name__}.{func.__name__} executed", payload=kwargs)
+        return func(*args, **kwargs)
+
+    return wrapper
+
+
+def send_custom_event(event: str = "", payload: Dict = {}):
+    """
+    This method can be called directly from anywhere in Haystack to send an event.
+    Enriches the given event with metadata and sends it to the posthog server if telemetry is enabled.
+    If telemetry has just been disabled, a final event is sent and the config file and the log file are deleted
+
+    :param event: Name of the event. Use a noun and a verb, e.g., "evaluation started", "component created"
+    :param payload: A dictionary containing event meta data, e.g., parameter settings
+    """
+    try:
+
+        def request_task(data):
+            event_properties = {**(NonPrivateParameters.apply_filter(data)), **_get_or_create_telemetry_meta_data()}
+            user_id = _get_or_create_user_id()
+            try:
+                posthog.capture(distinct_id=user_id, event=event, properties=event_properties)
+            except Exception:
+                pass
+            if is_telemetry_enabled() and is_telemetry_logging_to_file_enabled():
+                _write_event_to_telemetry_log_file(distinct_id=user_id, event=event, properties=event_properties)
+
+        def fire_and_forget(data):
+            Thread(target=request_task, args=(data,)).start()
```

Comment:
I agree that it would be difficult to extract and reuse these inner functions. However, by specifying them inside the scope `send_custom_event` I'd say it is made explicit that the inner functions should not be reused outside of `send_custom_event`. I understand that this limits our options to test the inner functions individually. For now, I'd say that the `test_send_event_via_decorator` test case covers all of the important functionality. Happy to discuss a refactoring of these tests and the inner functions in a separate issue though. 🙂